File: Leetcode/0347 Top K Frequent Elements.cpp
struct Pair{
   int num,freq;
   
   Pair(int n = 0,int f = 0) : num(n),freq(f){}
   
   bool operator<(const Pair& other) const {
       return freq < other.freq;  
   }
};

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map <int,int> freq;
        
        for(int num : nums)
            freq[num]++;

        priority_queue<Pair> pq;

        auto it = freq.begin();
        while(it != freq.end()){
            pq.push(Pair(it->first,it->second));
            it++;
        }

        vector<int> out;
        while(k-- && !pq.empty()){
            out.push_back(pq.top().num);
            pq.pop();
        }

        return out;
    }
};

File: Leetcode/0347 Top K Frequent Elements.py
class Solution:
    # O(n + n log k + k)
    def topKFrequent(self, nums, k):
        # O(n) time
        freq = Counter(nums)
        top_k_elems = []
        # O(n log k) time
        for e, f in freq.items():
            if len(top_k_elems) < k or f >= top_k_elems[0][0]:
                heappush(top_k_elems, (f, e))
                if len(top_k_elems) > k:
                    heappop(top_k_elems)
        # O(k) time
        return [e for _, e in top_k_elems]
